source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -21,76 +23,129 @@ public class MessageExchangeStream : IMessageExchangeStream
         readonly ILog log;
-        readonly StreamWriter streamWriter;
+        [Obsolete]
+        readonly StreamWriter streamWriter = null;
         readonly IMessageSerializer serializer;
-        readonly Version currentVersion = new Version(1, 0);
-        readonly ControlMessageReader controlMessageReader = new ControlMessageReader();
+        readonly Version currentVersion = new(1, 0);
+        readonly ControlMessageReader controlMessageReader = new();
 
-        public MessageExchangeStream(Stream stream, IMessageSerializer serializer, ILog log)
+        public MessageExchangeStream(Stream stream, IMessageSerializer serializer, AsyncHalibutFeature asyncHalibutFeature, ILog log)
         {
-            #if NETFRAMEWORK
-            this.stream = stream;
-            #else
-            this.stream = new RewindableBufferStream(stream, HalibutLimits.RewindableBufferStreamSize);
-            #endif
+#if NETFRAMEWORK
+            this.stream = asyncHalibutFeature.IsEnabled() ? new TimeoutStream(stream) : stream;
+#else
+            this.stream = asyncHalibutFeature.IsEnabled() ? new RewindableBufferStream(new TimeoutStream(stream), HalibutLimits.RewindableBufferStreamSize) : new RewindableBufferStream(stream, HalibutLimits.RewindableBufferStreamSize);
+#endif
             this.log = log;
-            streamWriter = new StreamWriter(this.stream, new UTF8Encoding(false)) { NewLine = "\r\n" };
             this.serializer = serializer;
+
+#pragma warning disable CS0612 // Type or member is obsolete
+            streamWriter = new StreamWriter(this.stream, new UTF8Encoding(false)) { NewLine = "\r\n" };
+#pragma warning restore CS0612 // Type or member is obsolete
+            
             SetNormalTimeouts();
         }
 
         static int streamCount;
 
+        [Obsolete]
         public void IdentifyAsClient()
         {
             log.Write(EventType.Diagnostic, "Identifying as a client");
             SendIdentityMessage($"{MxClient} {currentVersion}");
             ExpectServerIdentity();
         }
 
+        public async Task IdentifyAsClientAsync(CancellationToken cancellationToken)
+        {
+            log.Write(EventType.Diagnostic, "Identifying as a client");
+            await SendIdentityMessageAsync($"{MxClient} {currentVersion}", cancellationToken);
+            await ExpectServerIdentityAsync(cancellationToken);
+        }
+
+        [Obsolete]
         void SendControlMessage(string message)
         {
             streamWriter.WriteLine(message);
             streamWriter.Flush();
         }
 
+        [Obsolete]
         async Task SendControlMessageAsync(string message)
         {
             await streamWriter.WriteLineAsync(message).ConfigureAwait(false);
             await streamWriter.FlushAsync().ConfigureAwait(false);
         }
 
+        async Task SendControlMessageAsync(string message, CancellationToken cancellationToken)
+        {
+            await stream.WriteLineAsync(message, cancellationToken);
```

Comment:
`streamWrite` has been replaced with interacting directly with the `stream` for Async Halibut

Reply:
I think this should be "WriteControlMessage" since a line could end with `\n` or `\r\n` where control messages end with `\r\n`.